From aa8b6b480f10c2c205ba8c78d8e199df5b0808cc Mon Sep 17 00:00:00 2001
From: Xavier Boniface <xavierx.boniface@intel.com>
Date: Fri, 19 Apr 2013 15:03:45 +0200
Subject: [PORT FROM MAIN][BT] Add -testmode option to bdt to allow use of nohup

BZ: 101809

For some implementations (like with a TI chip), the Bluetooth chip is
reset when bdt closes. Thus, when bdt is launched from an adb shell,
it is not possible to unplug the USB cable (if the device needs to be
put in test mode and then placed in a shielded box for example).

A workaround for this is to call bdt with nohup. But because bdt takes
some commands from stdin, we need an option to place it directly in
test mode at startup. Hence this additional "-testmode" option.

Calling
  bdt -testmode

is equivalent to calling
  bdt
and then launch the following commands:
  enable
  <wait for enabling>
  dut_mode_configure 1

To call bdt in an adb shell and let it run even if the shell is
closed, call:
  nohup bdt -testmode &

Change-Id: Ib10e88870b79deae74edd615840ff1e0984c7a17
Orig-Change-Id: I789c3e96e1eea696dd51a23c6e3273e1f39f42ee
Category: aosp improvement
Domain: CWS.BT-Common
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Xavier Boniface <xavierx.boniface@intel.com>
Reviewed-on: http://android.intel.com:8080/104120
Reviewed-by: Gros, Julien <julien.gros@intel.com>
Tested-by: Zurmely, PierreX <pierrex.zurmely@intel.com>
---
 test/bluedroidtest/bluedroidtest.c |   28 ++++++++++++++++++++++++++++
 1 files changed, 28 insertions(+), 0 deletions(-)

diff --git a/test/bluedroidtest/bluedroidtest.c b/test/bluedroidtest/bluedroidtest.c
index 2b517df..7e09a2e 100644
--- a/test/bluedroidtest/bluedroidtest.c
+++ b/test/bluedroidtest/bluedroidtest.c
@@ -753,6 +753,34 @@ int main (int argc, char * argv[])
     /* Automatically perform the init */
     bdt_init();
 
+    /* Handling the '-testmode' option where we directly enable and put in test mode. */
+    if(argc == 2 && !strcmp(argv[1], "-testmode"))
+    {
+        bdt_log(":: Bluedroid test app called with '-testmode' option: put directly in test mode");
+
+        /* 1. enable bluetooth */
+        bdt_log(":: Bluedroid test app: -testmode: calling enable()");
+        bdt_enable();
+
+        /* 2. if bt was not enabled, wait for the adapter_state_changed() callback to be called
+         * and bt_enabled to be set to 1. */
+        bdt_log(":: Bluedroid test app: -testmode: waiting for enabling");
+        do
+        {
+            sleep(1);
+        }
+        while (!bt_enabled);
+
+        // Bluetooth is now enabled
+
+        /* 3. Set into test mode */
+        bdt_log(":: Bluedroid test app: -testmode: setting in test mode");
+        bdt_dut_mode_configure("1");
+
+        /* 4. Then wait, accepting further commands as normally */
+    }
+
+
     while(!main_done)
     {
         char line[128];
-- 
1.7.4.1

